### DISCUSSION TITLE
How to detect participant disconnection or leaving from room

### DISCUSSION THREAD

**Discussion by FarhodKurbonov** [ant-media/Ant-Media-Server#6254] [Q&A]
Short Description: Goal - Determine which of the conference participant left the room or not publishing own stream. There are situations when a participant’s mobile device may have low battery and can die or the conference application has been closed for some reason by user. 

Use case:
1. `User A` and `User B` joined the room, published/played local/remote streams
2. `User B` left the room for some reason (phone battery died, user killed the app etc)
Expected Result: `User A` should detect that `User B` is no longer publishing own stream and as a result it is not participant of the room  *immediately*. 

Additional Info:
 - Ant Media Server Version -  2.8.0
 - Conference implementation - MultiTrack Conference

It is fair to note possibility of room participants  by periodically (3, 5, 10 seconds) checking it via sending getRoomInfo request. Drawback of it next iteration of room information is binding to a timer that user has to set. Is there any other way to know if user is actively streaming or not on webRTC level? There were Server Side Notification such as `streamLeaved ` but it has been deleted due to cluster issue

**Comment by zerostack0** [comment]
Hi @FarhodKurbonov,

There are 2 callbacks that you can check. Whenever someone leaves the room or is dropped for some reason, you will receive a TRACK_LIST_UPDATED event which means that either someone joined or left. Then, you can call getBroadcastObject with room name. It returns the broadcast object from the broadcastObject info callback. You can check the subtracts to see the current participant list.
Please let us know if it works for you or not.

**Reply by FarhodKurbonov** [reply]
Hi @mustafaboleken thank you for you reply. Let us to try it. I will follow up about it for sure

**Reply by FarhodKurbonov** [reply]
Hi @mustafaboleken, we have started implementing reconnection across all our platforms (Android, iOS, Web) following your recommendation. We agree with @makhmudjon47. The behavior of `TRACK_LIST_UPDATED` is not meeting user expectations. We tested your implementation using the official [GitHub repository for the Ant Media Video Conferencing application](https://github.com/ant-media/conference-call-application/tree/main), both locally and on your deployed [Circle App](https://meet.antmedia.io/Conference/).

Here's the exact use case:

Steps to reproduce:

* `User A` joins the room (abc).
* `User B` joins the same room.
* `User A` turns off or physically unplugs his internet connection (WiFi/4G, etc.).

Actual Behavior:
When the internet is disconnected on `User A's` side, the app tries to reconnect and notifies `User A` which is perfect. Meanwhile, `User B` is unaware of what is happening on `User A's` side, and it appears as if `User A` is still online

Expected Behaviour:
When the internet is disconnected on `User A's` side, the app should notify `User B` about it. This is important because `User B` may continue speaking to `User A`, not realizing that hi is offline and trying to reconnect. The app should also display the reconnection status of `User A` on `User B's` side.

Additional Info:

Server Version 2.8.2
Tested version of Circle App: 2.8.2 and the version officially deployed on meet.antmedia.io.

We reviewed the AntMedia.js file in the Circle App and noticed that when a participant loses connectionthe server waits for them (does not close the websocket connection and session). However, `TRACK_LIST_UPDATED` is not triggered on the side of other participants. It only triggers when a client app (the one who has been having trouble with reconnection) drops the connection (closes a web browser tab or kills the app), which is predictable. In this scenario, the server closes the websocket connection and  session with the room participant and triggers `TRACK_LIST_UPDATED`.

Is there any way for server to notify room participants about  those ones who lost connection and tries to get connected?

**Comment by makhmudjon47** [comment]
@mustafaboleken  We're glad to hear that TRACK_LIST_UPDATED event works when people join or leave the room. We're also working on reconnection functionality.

However, we noticed that when someone disconnects while in the room, others don't receive the  TRACK_LIST_UPDATED event. We'd appreciate any insights on how to handle this situation for a smoother user experience.
